pennylane/optimize/qnspsa.py

```diff
@@ -0,0 +1,407 @@
+        blocking (boolean): when set to be True, the optimizer only accepts updates that lead to a
+            loss value no larger than the loss value before update, plus a tolerance. The tolerance
+            is set with the hyperparameter ``history_length``. The ``blocking`` option is
+            observed to help the optimizer to significantly converge faster
+        history_length (int): when ``blocking`` is True, the tolerance is set to be the average of
+            the cost values in the last ``history_length`` steps
+        seed (int): seed for the random sampling
+    """
+    # pylint: disable=too-many-arguments
+    # pylint: disable=too-many-instance-attributes
+    def __init__(
+        self,
+        stepsize=1e-3,
+        regularization=1e-3,
+        finite_diff_step=1e-2,
+        resamplings=1,
+        blocking=True,
+        history_length=5,
+        seed=None,
+    ):
+        self.stepsize = stepsize
+        self.reg = regularization
+        self.finite_diff_step = finite_diff_step
+        self.metric_tensor = None
+        self.k = 1
+        self.resamplings = resamplings
+        self.blocking = blocking
+        self.last_n_steps = np.zeros(history_length)
+        self.rng = np.random.default_rng(seed)
+
+    def step(self, cost, *args, **kwargs):
+        """Update trainable arguments with one step of the optimizer.
+
+        .. note::
+            When blocking is set to be True, ``step`` calls ``step_and_cost`` on the backend, as loss
+            measurements are required by the algorithm in this scenario.
+
+        Args:
+            cost (QNode): the QNode wrapper for the objective function for optimization
+            *args : variable length argument list for qnode
+            **kwargs : variable length of keyword arguments for the qnode
+
+        Returns:
+            np.array: the new variable values after step-wise update :math:`x^{(t+1)}`
+        """
+        if self.blocking:
+            warnings.warn(
+                "step_and_cost() instead of step() is called when "
+                "blocking is turned on, as the step-wise loss value "
+                "is required by the algorithm.",
+                stacklevel=2,
+            )
+            return self.step_and_cost(cost, *args, **kwargs)[0]
+
+        return self._step_core(cost, args, kwargs)
+
+    def step_and_cost(self, cost, *args, **kwargs):
+        r"""Update trainable parameters with one step of the optimizer and return
+        the corresponding objective function value after the step.
+
+        Args:
+            cost (QNode): the QNode wrapper for the objective function for optimization
+            *args : variable length argument list for qnode
+            **kwargs : variable length of keyword arguments for the qnode
+
+        Returns:
+            (np.array, float): the new variable values :math:`x^{(t+1)}` and the objective
+            function output prior to the step
```

Comment:
Correct me if I'm wrong.
```suggestion
            function output after the step
```

Reply:
I think it is the PennyLane's convention, that loss value is measured before the update. Here using gradient descent as an example:
https://pennylane.readthedocs.io/en/stable/code/api/pennylane.GradientDescentOptimizer.html

Reply:
Yep, we inherited this from Autograd. The reason is that, when you call `qml.grad()` with Autograd, there is _always_ an implicit forward pass before the backwards pass is performed (so 2 evaluations total).

So you get the forward value (before the optimization step) 'for free'.

If we supported the function output after the step, it would mean 3 total evaluations, which would scale very badly for end users. So we made the decision that if a user wants the loss after each step, they have to compute it explicitly themselves.